spyderlib/widgets/file_switcher.py

```diff
@@ -0,0 +1,401 @@
+        doc.documentLayout().draw(painter, ctx)
+
+        painter.restore()
+
+    def sizeHint(self, option, index):
+        options = QStyleOptionViewItemV4(option)
+        self.initStyleOption(options, index)
+
+        doc = QTextDocument()
+        doc.setHtml(options.text)
+        doc.setTextWidth(options.rect.width())
+        return QSize(doc.idealWidth(), doc.size().height())
+
+
+class UpDownFilter(QObject):
+    """Use with ``installEventFilter`` to get up/down arrow key press signal.
+    """
+    sig_up_down = Signal(int)
+
+    def eventFilter(self, src, e):
+        if e.type() == QEvent.KeyPress:
+            if e.key() == Qt.Key_Up:
+                self.sig_up_down.emit(-1)
+            elif e.key() == Qt.Key_Down:
+                self.sig_up_down.emit(+1)
+        return super(UpDownFilter, self).eventFilter(src, e)
+
+
+class FileSwitcher(QDialog):
+    sig_close_file = Signal(int)
+    sig_edit_file = Signal(int)
+    sig_edit_line = Signal(int)
+
+    def __init__(self, parent, tabs, tab_data):
+        QDialog.__init__(self, parent)
+
+        # Destroying the C++ object right after closing the dialog box,
+        # otherwise it may be garbage-collected in another QThread
+        # (e.g. the editor's analysis thread in Spyder), thus leading to
+        # a segmentation fault on UNIX or an application crash on Windows
+        self.setAttribute(Qt.WA_DeleteOnClose)
+
+        self.setWindowFlags(Qt.Dialog | Qt.FramelessWindowHint)
+        self.setWindowOpacity(0.9)
+        self.setModal(True)
+
+        self.tabs = tabs
+        self.tab_data = tab_data
+        self.filtered_index_to_path = None
+        self.full_index_to_path = None
+        self.path_to_line_count = None
+
+        edit_layout = QHBoxLayout()
+        self.edit = QLineEdit(self)
+        self.up_down_filter = UpDownFilter()
+        self.up_down_filter.sig_up_down.connect(self.handle_sig_up_down)
+        self.edit.installEventFilter(self.up_down_filter)
+        self.edit.returnPressed.connect(self.handle_sig_edit_file)
+        self.edit.textChanged.connect(lambda text: self.synchronize(None))
+
+        edit_layout.addWidget(self.edit)
+        self.button_help = HelperToolButton()
+        self.button_help.setIcon(get_std_icon('MessageBoxInformation'))
+        style = """
+            QToolButton {
+              border: 1px solid grey;
+              padding:0px;
+              border-radius: 2px;
+              background-color: qlineargradient(x1: 0, y1: 0, x2: 0, y2: 1,
+                  stop: 0 #f6f7fa, stop: 1 #dadbde);
+            }
+            """
+        self.button_help.setStyleSheet(style)
+        help_str = _("""
+            Press <b>Enter</b> to switch files.<br>
+            Or <b>Esc</b> to cancel.<br>
+            Type to filter filenames.<br>
+            Use <b>:number</b> to go to a line,<br>
+            e.g. 'main:42'.<br>
+            Press <b>Ctrl+W</b> to close current tab.
+            """)
```

Comment:
The tooltip seems a bit crowded, so maybe making lines longer, and adding double breaks <br><br> to separate paragraphs would make it better